pkg/supervisor/logwriter.go

```diff
@@ -0,0 +1,87 @@
+/*
+Copyright 2022 k0s authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package supervisor
+
+import (
+	"bytes"
+
+	"github.com/sirupsen/logrus"
+)
+
+// logWriter implements [io.Writer] by forwarding whole lines to log. In case
+// the lines get too long, it logs them in multiple chunks.
+//
+// This is in contrast to logrus's implementation of io.Writer, which simply
+// errors out if the log line gets longer than 64k.
+type logWriter struct {
+	log     logrus.FieldLogger // receives (possibly chunked) log lines
+	buf     []byte             // buffer in which to accumulate chunks; len(buf) determines the chunk length
+	len     int                // current buffer length
+	chunkNo uint               // current chunk number; 0 means "no chunk"
+}
+
+// Write implements [io.Writer].
+func (w *logWriter) Write(in []byte) (int, error) {
+	w.writeBytes(in)
+	return len(in), nil
+}
+
+func (w *logWriter) writeBytes(in []byte) {
+	// Fill and drain buffer with available data until everything has been consumed.
+	for rest := in; len(rest) > 0; {
+
+		n := copy(w.buf[w.len:], rest) // fill buffer with new input data
+		rest = rest[n:]                // strip copied input data
+		w.len += n                     // increase buffer length accordingly
+
+		// Loop over buffer as long as there are newlines in it
+		for off := 0; ; {
+			idx := bytes.IndexByte(w.buf[off:w.len], '\n')
```

Comment:
Maybe: 

```suggestion
			idx := bytes.IndexRune(w.buf[off:w.len], '\n')
```

It could theoretically break a multibyte character that contains 0x0a when using IndexByte

Reply:
True. Better use unicode-aware stuff!

Reply:
OTOH, I think we assume UTF-8 or some single-byte ASCII superset here. IIUC, there's no valid UTF-8 byte sequence that may contain any 0x0a which is actually _not_ a line break. ASCII characters are encoded as single bytes as-is, everything else as byte sequence where the highest bit is set on all bytes, i.e. they're never in the ASCII plane.

Things would get quite complicated if we'd want to add proper charset support here. WDYT?

Reply:
... but chunking could actually break a multibyte character.... :thinking: